### ISSUE TITLE
[SSH-7408] Incorrect port detection when sshd is socket activated

### ISSUE THREAD

**Issue by chr0mag** [CISOfy/lynis#629]
**Describe the bug**
Lynis is unable to detect the correct port when _sshd_ is started via _systemd_ socket activation.

I use _systemd_'s [sshd.socket](https://wiki.archlinux.org/index.php/OpenSSH#Daemon_management) to start OpenSSH. To change the listen ip/port for the socket, one has to customize the _systemd_ unit file as shown below. This overrides the _Port_ & _ListenAddress_ settings configured in _sshd_config_ so even when the _sshd -T_ output shows port 22 this is not, in fact the port on which _sshd.socket_ is listening. The port in this case discovered by inspecting the _ListenStream_ option in the relevant unit file.

```
# /usr/lib/systemd/system/sshd.socket
[Unit]
Conflicts=sshd.service
Wants=sshdgenkeys.service

[Socket]
ListenStream=22
Accept=yes

[Install]
WantedBy=sockets.target

# /etc/systemd/system/sshd.socket.d/override.conf
[Socket]
ListenStream=
ListenStream=10.14.14.1:11711
FreeBind=true
```

**Version**
 - Distribution: Archlinux
 - Lynis v2.7

**Expected behavior**
Correct port is detected.

**Output**
```
2019-01-28 21:58:56 ===---------------------------------------------------------------===
2019-01-28 21:58:56 Performing test ID SSH-7408 (Check SSH specific defined options)
2019-01-28 21:58:56 Test: Checking specific defined options in /tmp/lynis.6KCCimwHW3
...
2019-01-28 21:58:58 Test: Checking Port in /tmp/lynis.6KCCimwHW3
2019-01-28 21:58:58 Result: Option Port found
2019-01-28 21:58:58 Result: Option Port value is 22
2019-01-28 21:58:58 Result: SSH option Port is in a weak configuration state and should be fixed
2019-01-28 21:58:58 Suggestion: Consider hardening SSH configuration [test:SSH-7408] [details:Port (22 --> )] [solution:-]
2019-01-28 21:58:58 Hardening: assigned partial number of hardening points (0 of 3). Currently having 134 points (out of 194)
...
```

**Additional context**
_ss --listening --numeric --tcp --processes_
`LISTEN              0                   128                                 10.14.14.1:11711                                0.0.0.0:*                  users:(("systemd",pid=1,fd=46))`

It is actually _systemd_ that is listening on the socket and only spawning an _sshd_ instance when a client connects. More details [here](http://0pointer.de/blog/projects/inetd.html).

**Comment by mboelen** [comment]
Thanks for reporting. This seems to be an edge case. Can you share your rationale behind this setup? Then based on that we can make a decision if and what to change.

**Comment by chr0mag** [comment]
Definitely an edge case and not high priority, but probably becoming more common since its the recommended way to run `sshd` in most cases according to the _systemd_ folks. 

The [link](http://0pointer.de/blog/projects/inetd.html) in my original post goes into detail, but as I understand it socket-activating `sshd` provides better client isolation - each connection to the socket spawns its own process - which also makes it more robust since anything affecting a specific connection won't affect the listening socket (and therefore won't affect new connection attempts) or other active connections. Socket activation also provides more efficient resource utilization for most server cases where a remote login happens infrequently since running the socket is basically free until a connection happens, compared to `sshd.service` which is a process that is always running.

**Comment by mboelen** [comment]
Gave this some thought. As you discovered, Lynis gathers this information from the SSH daemon. Even if another service is hijacking the actual listening port, it is still SSH being configured in a particular way. The resources that OpenSSH uses are on most systems very low, so I doubt people will specifically use it for that way. The isolation might help. At the same time, the complexity of the system increases.

In conclusion: low priority, nice-to-have. Will close this request for now, to keep the open queue as low as possible. Feel free to submit a pull request if you got the time.
